controllers/etcd_controller.go

```diff
@@ -1077,8 +1095,5 @@ func (r *EtcdReconciler) SetupWithManager(mgr ctrl.Manager, workers int, ignoreO
 	builder = builder.WithEventFilter(druidpredicates.Or(predicates...))
 	return builder.
 		For(&druidv1alpha1.Etcd{}).
-		Owns(&appsv1.StatefulSet{}).
```

Comment:
Shouldn't you add these `Owns()` calls if the operation annotation is ignored?

Reply:
Fine for me. In the Gardener context it's only important that these mapping to the `Etcd` resource are not happening as they bypass the predicates.

Reply:
I disagree. I don't think removing `Owns` on managed downstream resources (`StatefulSet`, `ConfigMap` and `Service`) is a good idea. Even if operation annotation is not ignored. It is the job of druid, as an etcd controller, to reconcile the `StatefulSet` if it is modified incompatibly and/or deleted, no?

Please do not see druid as only a gardener extension. It is also a controller that is in charge of managing etcd. I am increasingly regretting the decision not to introduce a separate etcd extension controller in addition to druid which ideally should manage only the etcd.

Reply:
As said, the mappings introduced by `Owns()` function are bypassing the predicates. This does not fit to the Gardener approach. Also, in the Gardener context it is not necessary that it reconciles on `StatefulSet`/`ConfigMap`/`Service` changes. These resources live in the seed and no end-user tampers with it. Yes, this is not the mainstream controller convention, but it is what we need to ensure shoot reconciliations only happen in the maintenance time window, and especially only then when Gardener demands a reconciliation.

As said above - I'm fine if you add these `Owns()` calls when the operation annotation is ignored. When druid is used in Gardener this will anyways never be the situation, hence, I don't care.

From my point of of view you have two options now:

1. Remove the `Owns()` completely or add them conditionally if operation annotation is ignored.
2. Disagree with 1), go one step back and develop something that fits to the nature of Gardener. 

Druid is blocking a Gardener + extensions release for 2 weeks now. Please decide - either we get a first working druid version in now and you can improve the code in the iterations later on, or we don't do it. Then I'll go ahead and revert everything that is druid-related for now so that Gardener can be released. Please tell me how to proceed.

Reply:
Going with option 1 will get gardener unblocked. @amshuman-kr, I'll get it out this way and then ideate and work on the owns() functionality with druid. wdyt?

Reply:
I am uncomfortable with 1. But even otherwise, considering the number of issues we have faced with the integration, how about the following choices?

1. Remove the Owns() completely or add them conditionally if operation annotation is ignored. But  put the whole etcd-druid integration under a feature gate and disable the feature gate in gardener in this release until we have resolved all the issues currently understood and added tests.
1. Revert the current etcd-druid integration, go one step back and develop something that fits to the nature of Gardener and then file a fresh integration.

Reply:
I don't get it, we are working since two weeks on resolving the issues and are getting closer now. This PR is necessary for sure, then we will run another test and check if etcd-druid behaves correctly now. Only then we will release Gardener. Why do we now start talking about introducing a feature gate? It doesn't make sense at this point anymore. Also, we would have to touch all extensions as well (and they don't have feature gates yet). Then we can also go ahead and revert everything and wait for a cleaner version of druid.

I am open to try getting the druid ready for 1-2 more days, but as said, this PR would be needed for sure. So, please decide how you want to continue.

PS: I'm also somewhat confused why you said this:

>  I am increasingly regretting the decision not to introduce a separate etcd extension controller in addition to druid which ideally should manage only the etcd.

My understanding was that etcd-druid was MAINLY designed to fit to Gardener (as it's also developed in the `gardener` org). If this can be made optional to also work outside of Gardener environment then fine, no worries, but the prio 1 should be Gardener. Why else should we develop an etcd operator if we don't use it in Gardener?

Reply:
If we remove owns{}, it gets implemented how gardener was already doing it. if it goes in this way, now. it wouldnt break the existing gardener contract on the controlplane.
Next iteration I can add in the owns logic that is acceptable with gardener contract. That way release is unblocked too. I feel putting the feature under feature-gate now is similar to reverting effortwise.

Reply:
@rfranzke I just aired my overall concern. I am just worried what else we have missed out.

> PS: I'm also somewhat confused why you said this:

>> I am increasingly regretting the decision not to introduce a separate etcd extension controller in addition to druid which ideally should manage only the etcd.

Sorry for not giving the background. Initially there was some discussion about whether just druid will be enough or we will need an additional gardener extension controller that would take care of gardener-specific stuff. We ended up deciding against the additional extension controller. I am not sure if an additional extension controller would have solved this particular problem.

My main concern with "Remove the Owns() completely or add them conditionally if operation annotation is ignored." is that earlier without druid, if something went wrong with etcd statefulset, gardener/gardenlet would reconcile it. Now with druid removing `Owns()`, nobody would unless it is maintenance window or forced reconciliation. Is this correct understanding? Am I missing something?

Reply:
> My understanding was that etcd-druid was MAINLY designed to fit to Gardener (as it's also developed in the gardener org). If this can be made optional to also work outside of Gardener environment then fine, no worries, but the prio 1 should be Gardener. Why else should we develop an etcd operator if we don't use it in Gardener?

To be clear, my concern is not gardener-specific stuff in druid. That is fine. My concern is the correct behaviour if something goes wrong with the statefulset etc.Please see [above](https://github.com/gardener/etcd-druid/pull/34#discussion_r400739260).

Reply:
> is that earlier without druid, if something went wrong with etcd statefulset, gardener/gardenlet would reconcile it.

Gardener would have sent its desired state of the StatefulSet again - but only if the shoot reconciliation starts. If the statefulset fails in between then Gardener does nothing (though, this  never happened and basically also cannot happen unless a human intervenes).

> Now with druid removing Owns(), nobody would unless it is maintenance window or forced reconciliation. Is this correct understanding?

Exactly, as described, this is exactly how it was before. When Gardener decides to reconcile a shoot then it annotates the `Etcd` with the operation annotation which tells druid to start. If this annotation does not exist then druid shall not act (because Gardener didn't demand it). With `Owns()` druid would reconcile always if there is for example a status update in the statefulset. It wouldn't respect the predicates (operation annotation), leading to reconciliations outside of the maintenance time window and eventually to disruptions.

Reply:
> If the statefulset fails in between then Gardener does nothing (though, this never happened and basically also cannot happen unless a human intervenes).

If I understood right then if statefulset is unhealthy then even Gardner would have done nothing until reconciliation is triggered (manually or in maintenance window). This addresses my concern. Then I am fine going with "1. Remove the Owns() completely or add them conditionally if operation annotation is ignored."

Reply:
If i understand right, we need `ignoreOperationAnnotation` so that etcd-druid's reconciliation can be controlled by gardener by adding/removing the "reconcile" annotation during maintenance window.
Will it be cleaner/easier if we instead add a `reconciliationTimeWindow` itself under `etcd`'s spec. etcd-controller would then reconcile based only on the field if set, otherwise all the time.